### PR TITLE
keepalived: update version to 2.0.18

### PR DESCRIPTION
**Pull request by feckert**
Maintainer: me
Compile tested: lantiq_xrx200 and x86_64, APU3, OpenWrt master
Run tested: x86_64, APU3, Openwrt master, start with default config

Description:
Update version to 2.0.18
Fix shell check warnings

### CODE REVIEW

**Comment by neheb** [comment]
Can you look at running the init script through shellcheck? I see a lot of warning being thrown.

**Comment by feckert** [comment]
@neheb Thanks for the hint.
I have fixed the most shellcheck warnings, but some could not be solved.

> In net/keepalived/files/keepalived.init line 48:
> 		if [ "${opt:0:7}" = "no_val_" ]; then
>                       ^-- SC2039: In POSIX sh, string indexing is undefined.

This works as expected on ash so i do not want to change this

> In net/keepalived/files/keepalived.init line 49:
> 			opt="${opt:7}"
>                              ^-- SC2039: In POSIX sh, string indexing is undefined.

This works as expected on ash so i do not want to change this

> In net/keepalived/files/keepalived.init line 93:
> 		notify="$(echo "$notify" | tr 'a-z' 'A-Z')"
>                                               ^-- SC2018: Use '[:lower:]' to support accents and foreign alphabets.
>                                                     ^-- SC2019: Use '[:upper:]' to support accents and foreign alphabets.
> 

I have tried but the suggestion does not work

> In net/keepalived/files/keepalived.init line 99:
> 	local linkbeat_use_polling notification_email
>         ^-- SC2039: In POSIX sh, 'local' is undefined.
>                                    ^-- SC2034: notification_email appears unused. Verify it or export it.

This is wrong I am using this variable so I think shellcheck  does do some thing wrong

> In net/keepalived/files/keepalived.init line 290:
> 		printf '%bauthentication {\n' "${INDENT_1}" >> "$KEEPALIVED_CONF"
>                 ^-- SC2129: Consider using { cmd1; cmd2; } >> file instead of individual redirects.

The suggestion does not work 

> In net/keepalived/files/keepalived.init line 321:
> 	for opt in virtual_routes; do
>                    ^-- SC2043: This loop will only run once, with opt='virtual_routes'.
> In net/keepalived/files/keepalived.init line 333:
> 	for opt in track_script; do
>                    ^-- SC2043: This loop will only run once, with opt='track_script'.
> In net/keepalived/files/keepalived.init line 345:
> 	for opt in track_interface; do
>                    ^-- SC2043: This loop will only run once, with opt='track_interface'.
> In net/keepalived/files/keepalived.init line 357:
> 	for opt in unicast_peer; do
>                    ^-- SC2043: This loop will only run once, with opt='unicast_peer'.

I know that's wrong, but for the sake of clarity it looks better.

> In net/keepalived/files/keepalived.init line 391:
> 		printf '%burl {\n' "${INDENT_3}" >> "$KEEPALIVED_CONF"
>                 ^-- SC2129: Consider using { cmd1; cmd2; } >> file instead of individual redirects.

This does not work

**Comment by neheb** [comment]
I usually set the shell to bash, fix warnings, and then move to dash. BusyBox Ash falls in between the two in terms of capability. I'll take a look to see why some of those warnings don't work.

File: net/keepalived/files/keepalived.init [hunk 1]
```diff
@@ -379,17 +381,17 @@ url() {
...
-	[ -n "$digest" -a -n "$path" ] && {
-		printf "${INDENT_3}url {\n" >> $KEEPALIVED_CONF
-		printf "${INDENT_4}path "$path"\n" >> $KEEPALIVED_CONF
-		printf "${INDENT_4}digest $digest\n" >> $KEEPALIVED_CONF
-		printf "${INDENT_3}}\n" >> $KEEPALIVED_CONF
+	[ -n "$digest" ] && [ -n "$path" ] && {
+		printf '%burl {\n' "${INDENT_3}" >> "$KEEPALIVED_CONF"
+		printf '%bpath %s\n' "${INDENT_4}" "$path" >> "$KEEPALIVED_CONF"
+		printf '%bdigest %s\n' "${INDENT_4}" "$digest" >> "$KEEPALIVED_CONF"
+		printf '%b}\n' >> "${INDENT_3}" "$KEEPALIVED_CONF"
```
Comment (neheb) [inline net/keepalived/files/keepalived.init]: This is a typo.

**Comment by neheb** [comment]
I looked at SC2129. It might not be compatible with the specific syntax that it matches on. Changed to if/then. Maybe that works.

It also seems there are more opportunities to group the writes. I made a patch for it:

```
--- a/net/keepalived/files/keepalived.init
+++ b/net/keepalived/files/keepalived.init
@@ -17,9 +17,11 @@ config_section_open() {
 	local tag="$1"
 	local name="$2"
 
-	printf '%s' "$tag" >> "$KEEPALIVED_CONF"
-	[ -n "$name" ] && printf ' %s' "$name" >> "$KEEPALIVED_CONF"
-	printf ' {\n' >> "$KEEPALIVED_CONF"
+	{
+		printf '%s' "$tag"
+		[ -n "$name" ] && printf ' %s' "$name"
+		printf ' {\n'
+	} >> "$KEEPALIVED_CONF"
 }
 
 config_section_close() {
@@ -52,17 +54,19 @@ print_elems_indent() {
 		config_get "$opt" "$config" "$opt"
 		eval optval=\$"$opt"
 		[ -z "$optval" ] && continue
-		printf '%b%s' "$indent" "$opt" >> "$KEEPALIVED_CONF"
-		[ "$no_val" = "0" ] && {
-			local words=0
-			words="$(echo "$optval" | wc -w)"
-			if [ "$words" -gt 1 ]; then
-				printf ' "%s"' "$optval" >> "$KEEPALIVED_CONF"
-			else
-				printf ' %s' "$optval" >> "$KEEPALIVED_CONF"
-			fi
-		}
-		printf '\n' >> "$KEEPALIVED_CONF"
+		{
+			printf '%b%s' "$indent" "$opt"
+			[ "$no_val" = "0" ] && {
+				local words=0
+				words="$(echo "$optval" | wc -w)"
+				if [ "$words" -gt 1 ]; then
+					printf ' "%s"' "$optval"
+				else
+					printf ' %s' "$optval"
+				fi
+			}
+			printf '\n'
+		} >> "$KEEPALIVED_CONF"
 	done
 	unset optval
 }
@@ -76,11 +80,13 @@ print_list_indent() {
 	eval lst_elems=\$"$lst"
 	[ -z "$lst_elems" ] && return 0
 
-	printf '%b%s {\n' "$indent" "$lst" >> "$KEEPALIVED_CONF"
-	for e in $lst_elems; do
-		printf '%b%s\n' "${indent}${INDENT_1}" "$e">> "$KEEPALIVED_CONF"
-	done
-	printf '%b}\n' "$indent" >> "$KEEPALIVED_CONF"
+	{
+		printf '%b%s {\n' "$indent" "$lst"
+		for e in $lst_elems; do
+			printf '%b%s\n' "${indent}${INDENT_1}" "$e"
+		done
+		printf '%b}\n' "$indent"
+	} >> "$KEEPALIVED_CONF"
 }
 
 print_notify() {
@@ -88,11 +94,14 @@ print_notify() {
 	shift
 	local name="$1"
 	shift
+
+	{
 	for notify in "$@"; do
-		printf '%b%s' "${INDENT_1}" "$notify">> "$KEEPALIVED_CONF"
+		printf '%b%s' "${INDENT_1}" "$notify"
 		notify="$(echo "$notify" | tr 'a-z' 'A-Z')"
-		printf ' "/bin/busybox env -i ACTION=%s TYPE=%s NAME=%s /sbin/hotplug-call keepalived"\n' "$notify" "$type" "$name" >> "$KEEPALIVED_CONF"
+		printf ' "/bin/busybox env -i ACTION=%s TYPE=%s NAME=%s /sbin/hotplug-call keepalived"\n' "$notify" "$type" "$name"
 	done
+	} >> "$KEEPALIVED_CONF"
 }
 
 global_defs() {
@@ -136,16 +145,18 @@ print_ipaddress_indent() {
 	# If no address exit
 	[ -z "$address" ] && return 0
 
-	if [ -z "$device" ]; then
-		printf '%b%s' "$indent" "$address" >> "$KEEPALIVED_CONF"
-	else
-		# Add IP address/netmask and device
-		printf '%b%s dev %s' "$indent" "$address" "$device">> "$KEEPALIVED_CONF"
-		# Add scope
-		[ -n "$scope" ] && printf ' scope %s' "$scope" >> "$KEEPALIVED_CONF"
-	fi
+	{
+		if [ -z "$device" ]; then
+			printf '%b%s' "$indent" "$address"
+		else
+			# Add IP address/netmask and device
+			printf '%b%s dev %s' "$indent" "$address" "$device"
+			# Add scope
+			[ -n "$scope" ] && printf ' scope %s' "$scope"
+		fi
 
-	printf '\n' >> "$KEEPALIVED_CONF"
+		printf '\n'
+	} >> "$KEEPALIVED_CONF"
 }
 
 static_ipaddress() {
@@ -179,27 +190,28 @@ print_route_indent() {
 	# Default indent
 	[ -z "$indent" ] && indent="$INDENT_1"
 
-	[ "$blackhole" -gt 0 ] && {
-		printf '%bblackhole %s\n' "$indent" "$address" >> "$KEEPALIVED_CONF"
-		return 0
-	}
-	# Add src addr or address
-	if [ -n "$src_addr" ]; then
-		printf '%bsrc %s %s' "$indent" "$src_addr" "$address" >> "$KEEPALIVED_CONF"
-	else
-		[ -z "$device" ] && return 0
-		printf '%b%s' "$indent" "$address" >> "$KEEPALIVED_CONF"
-	fi
-	# Add route/gateway
-	[ -n "$gateway" ] && printf ' via %s' "$gateway" >> "$KEEPALIVED_CONF"
-	# Add device
-	printf ' dev %s' "$device" >> "$KEEPALIVED_CONF"
-	# Add scope
-	[ -n "$scope" ] && printf ' scope %s' "$scope" >> "$KEEPALIVED_CONF"
-	# Add table
-	[ -n "$table" ] && printf ' table %s' "$table" >> "$KEEPALIVED_CONF"
-	printf '\n' >> "$KEEPALIVED_CONF"
-
+	{
+		[ "$blackhole" -gt 0 ] && {
+			printf '%bblackhole %s\n' "$indent" "$address"
+			return 0
+		}
+		# Add src addr or address
+		if [ -n "$src_addr" ]; then
+			printf '%bsrc %s %s' "$indent" "$src_addr" "$address"
+		else
+			[ -z "$device" ] && return 0
+			printf '%b%s' "$indent" "$address"
+		fi
+		# Add route/gateway
+		[ -n "$gateway" ] && printf ' via %s' "$gateway"
+		# Add device
+		printf ' dev %s' "$device"
+		# Add scope
+		[ -n "$scope" ] && printf ' scope %s' "$scope"
+		# Add table
+		[ -n "$table" ] && printf ' table %s' "$table"
+		printf '\n'
+	} >> "$KEEPALIVED_CONF"
 }
 
 print_track_elem_indent() {
@@ -216,9 +228,11 @@ print_track_elem_indent() {
 
 	[ -z "$value" ] && return 0
 
-	printf '%b%s' "$indent" "$value" >> "$KEEPALIVED_CONF"
-	[ -n "$weight" ] && printf ' weight %s' "$weight" >> "$KEEPALIVED_CONF"
-	printf '\n' >> "$KEEPALIVED_CONF"
+	{
+		printf '%b%s' "$indent" "$value"
+		[ -n "$weight" ] && printf ' weight %s' "$weight"
+		printf '\n'
+	} >> "$KEEPALIVED_CONF"
 }
 
 static_routes() {
@@ -256,7 +270,7 @@ vrrp_sync_group() {
 	config_get group "$1" group
 	[ -z "$group" ] && return 0
 
-	# Check if we have 'vrrp_instance's defined for 
+	# Check if we have 'vrrp_instance's defined for
 	# each member and remove names with not vrrp_instance defined
 	for m in $group; do
 		vrrp_instance_check "$m" && valid_group="$valid_group $m"
@@ -286,12 +300,14 @@ vrrp_instance() {
 
 	config_get auth_type "$1" auth_type
 	config_get auth_pass "$1" auth_pass
-	[ -n "$auth_type" ] && [ -n "$auth_pass" ] && {
-		printf '%bauthentication {\n' "${INDENT_1}" >> "$KEEPALIVED_CONF"
-		printf '%bauth_type %s\n' "${INDENT_2}" "$auth_type" >> "$KEEPALIVED_CONF"
-		printf '%bauth_pass %s\n' "${INDENT_2}" "$auth_pass" >> "$KEEPALIVED_CONF"
-		printf '%b}\n' "${INDENT_1}" >> "$KEEPALIVED_CONF"
-	}
+	if [ -n "$auth_type" ] && [ -n "$auth_pass" ];then
+	{
+		printf '%bauthentication {\n' "${INDENT_1}"
+		printf '%bauth_type %s\n' "${INDENT_2}" "$auth_type"
+		printf '%bauth_pass %s\n' "${INDENT_2}" "$auth_pass"
+		printf '%b}\n' "${INDENT_1}"
+	} >> "$KEEPALIVED_CONF"
+	fi
 
 	print_elems_indent "$1" "$INDENT_1" state interface \
 		mcast_src_ip unicast_src_ip virtual_router_id version priority \
@@ -387,12 +403,14 @@ url() {
 	config_get path "$1" path
 	config_get digest "$1" digest
 
-	[ -n "$digest" ] && [ -n "$path" ] && {
-		printf '%burl {\n' "${INDENT_3}" >> "$KEEPALIVED_CONF"
-		printf '%bpath %s\n' "${INDENT_4}" "$path" >> "$KEEPALIVED_CONF"
-		printf '%bdigest %s\n' "${INDENT_4}" "$digest" >> "$KEEPALIVED_CONF"
-		printf '%b}\n' >> "${INDENT_3}" "$KEEPALIVED_CONF"
-	}
+	if [ -n "$digest" ] && [ -n "$path" ];then
+	{
+		printf '%burl {\n' "${INDENT_3}"
+		printf '%bpath %s\n' "${INDENT_4}" "$path"
+		printf '%bdigest %s\n' "${INDENT_4}" "$digest"
+		printf '%b}\n' "${INDENT_3}"
+	} >> "$KEEPALIVED_CONF"
+	fi
 }
 
 url_list() {
```

**Comment by feckert** [comment]
@neheb I am not sure if we should do this. I have seen that the debugging get more difficult as it should. This is because every thing get redirected to the file if I do an echo. I think this is only a performance issue, but the configuration will generated only on startup. So I think we should leave them as it is now

File: net/keepalived/files/keepalived.init (hunk 1 above)
Comment (feckert) [reply net/keepalived/files/keepalived.init]: Fixed thanks :+1:

**Comment by feckert** [comment]
@neheb could I merge this from your side?

**Comment by neheb** [comment]
Go for it.

**Pull request merged**